### PR TITLE
DP-1946

### PR DESCRIPTION
**Pull request by michaelsmithhx**
GBQ integration and includes a fix for optional arrays so that items like user tags generate correctly

### CODE REVIEW

File: src/test.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      "type": "string"
+    },
+    "title": {
+      "type": "string"
+    },
+    "address": {
+      "type": "object",
+      "properties": {
+        "street_address": {
+                    "type": "string"
```
Comment (hx-markterry) [inline src/test.json:20]: indentation is wrong here

File: README.md [hunk 2]
```diff
@@ -5,29 +5,22 @@
...
+
+Note that some features involve bespoke interpretation of schema details suited to our environment.
+
 ## Install
 
     npm install jsonschema-bigquery
 
 ## Consume
 
-    const jsonSchemaBigquery = require('jsonschema-bigquery')
```
Comment (hx-markterry) [inline README.md:14]: Can you keep these code blocks as i see the modules primary function is to allow json schema to converted programmatically. The JSBQ command is an added extra.

File: package.json [hunk 3]
```diff
@@ -30,8 +31,17 @@
   },
   "homepage": "https://github.com/thedumbterminal/jsonschema-bigquery#readme",
   "devDependencies": {
+    "eslint": "5.6.1",
```
Comment (hx-markterry) [inline package.json:34]: Can you add `^` to the versions please.

File: package.json [hunk 4]
```diff
@@ -4,7 +4,8 @@
   "description": "Convert JSON schema to Google BigQuery schema",
   "main": "src/index.js",
   "scripts": {
-    "test": "mocha test"
+    "test": "mocha test",
+    "lint": "eslint ."
```
Comment (hx-markterry) [inline package.json:8]: This should of been done in a separate PR.

Don't forget this is not a HX project so changing/applying linting on someone else project is a bit intrusive.

File: src/index.js [hunk 5]
```diff
@@ -1,104 +1,118 @@
...
+  return  {
+    schema: {
+      fields: fields
+    }
+  }
 }
 
 jsonSchemaBigquery._isNested = (schema) => schema.type === 'object'
-jsonSchemaBigquery._isRepeated = (schema) => schema.type === 'array'
+jsonSchemaBigquery._isRepeated = (schema) => schema.type === 'array' || (Array.isArray(schema.type) && schema.type.includes('array'))
```
Comment (hx-markterry) [inline src/index.js]: Do we have any schemas where we have array of types, which include arrays?

File: src/jsbq.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  logger.info(JSON.stringify(tableOptions))
+  
+  logger.info('Extracting table name from schema...')
+  const tableName = util.get_table_id(jsonSchema.id)
+  logger.info('Table name is ' + tableName)
+
+  logger.info('Setting table options...')
+  tableOptions.friendly_name = jsonSchema.title
+  tableOptions.description = jsonSchema.description || jsonSchema.title
+  tableOptions.timePartitioning = {
```
Comment (hx-markterry) [inline src/jsbq.js:21]: We need to mention this behaviour in the docs as other users will need to be aware of it.

File: test/samples/allOf/input.json [hunk 7]
```diff
@@ -1,27 +1,31 @@
...
-							"type": "string"
-						}
-					}
-				}
-			]
-		}
-	}
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/allOf/input.json:4]: The next example covers this already, don't need to add allof here.

File: test/samples/anyOfMultipleTypes/input.json [hunk 8]
```diff
@@ -1,20 +1,24 @@
...
-			]
-		}
-	},
-	"required": [
-		"first_name"
-	]
-}	
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/anyOfMultipleTypes/input.json:4]: The example above covers this already, don't need to add allof here.

File: test/samples/complex/input.json [hunk 9]
```diff
@@ -1,18 +1,22 @@
...
-				},
-				"country": {
-					"type" : "string"
-				}
-			}
-		}
-	}
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Complex example",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/complex/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: test/samples/fieldDescription/input.json [hunk 10]
```diff
@@ -1,11 +1,15 @@
...
-	"type": "object",
-	"properties": {
-		"first_name": {
-			"description": "the first name",
-			"type": "string"
-		}
-	}
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/fieldDescription/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: test/samples/nestedRepeated/input.json [hunk 11]
```diff
@@ -1,18 +1,22 @@
...
-					"first": {
-						"type": "string"
-					}
-				}
-			}
-		}
-	}
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/nestedRepeated/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: test/samples/nullable/input.json [hunk 12]
```diff
@@ -1,16 +1,20 @@
...
-		"last_name": {
-			"type": "string"
-		}
-	},
-	"required": [
-		"first_name"
-	]
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/nullable/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: test/samples/repeated/input.json [hunk 13]
```diff
@@ -1,13 +1,17 @@
...
-		"names": {
-			"type": "array",
-			"items": {
-				"type": "string"
-			}
-		}
-	}
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/repeated/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: test/samples/simple/input.json [hunk 14]
```diff
@@ -1,13 +1,17 @@
...
-		"first_name": {
-			"type": "string"
-		},
-		"last_name": {
-			"type": "string"
-		}
-	}
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/simple/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: test/samples/timestamp/input.json [hunk 15]
```diff
@@ -1,11 +1,15 @@
...
-	"type": "object",
-	"properties": {
-		"modified": {
-			"type": "string",
-			"format": "date-time"
-		}
-	}
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/timestamp/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: test/samples/typeArray/input.json [hunk 16]
```diff
@@ -1,22 +1,26 @@
...
-				"null"
-			]
-		}
-	},
-	"required": [
-		"first_name"
-	]
+    "id": "http://yourdomain.com/schemas/myschema.json",
+    "description": "Example description",
+    "allOf": [
```
Comment (hx-markterry) [inline test/samples/typeArray/input.json:4]: We dont need to add allOf here as we have other examples that do this, can we revert this please.

File: README.md [hunk 17]
```diff
@@ -5,29 +5,22 @@
...
 Convert JSON schema to Google BigQuery schema
 
+This includes the ability to:
+
+(1) Create tables
+(2) Patch tables
+
+Further enhancements are planned: delete tables (dev only), create dataset, set data ACLs
+
+Note that some features involve bespoke interpretation of schema details suited to our environment.
```
Comment (hx-markterry) [inline README.md:15]: Can we list any specifics here?

**Review by hx-markterry — CHANGES_REQUESTED** [review]
Nice work on getting the allOf working, only a few comments and need to revert most of the example changes.

**Comment by michaelsmithhx** [comment]
Closing as new PR will introduce the cloned bigjson.py logic

**Pull request closed without merging**